Recursive yet unboxed

rlepigre opened the pull request:
We (@gasche and I) give a new implementation of the unboxability check for single-constructor types, following theoretical investigations presented in our [JFLA 2019 paper](https://arxiv.org/abs/1811.02300) (which refines the work of @SimonColin as part of an internship with @gasche). Mutually-recursive type definitions can now contain unboxed declarations, which was not the case before.

Commit:
typedecl_separability: support for algebraic datatypes (non-GADTs)

Commit:
Cleaned up [@@unbox] validity check.

gasche committed:
typedecl_separability: destructure type definition to access unboxing information

In the future, we could move the arity-related checks
(one constructor, one parameter) and error logic from typedecl.ml to
typedecl_separability.ml.

gasche committed:
add separability signatures in type declarations

(this changes the .cmi format and thus requires a bootstrap,
to follow as a separate commit)

(includes bug fixes by Rodolphe Lepigre)

Commit:
typedecl_separability: expose the generic Typedecl_properties interface

Commit:
Update changelog.

gasche commented:
The history is not as clean as I would like, but we believe that the code is ready for review. I hope to rebase the PR later and clean the history in passing.

Below are two remarks that I wrote on the implementation.

.

We distinguish the notion of "unboxability" (of type definitions) and
"separability", with respect to the Config.flat_float_array setting:
separability is a semantic notion that is always well-defined
(and does not depend on `Config.flat_float_array`), but "unboxability"
only depends on separability when `Config.flat_float_array` is set.

Ideally, the typedecl_separability module would never depend on
`Config.flat_float_array`, and this flag would be tested outside, by the
caller. In practice, it is more regular (with respect to other
`typedecl_*` modules) to have the check in the module, but it is only at
the edge, in the compute_decl function that is exposed outside
(and also in the `Types.Separability.default_signature` function).

.

There is a strange duplication between `Types.Separability.default_signature`
and `Types_separability.msig_of_external_type`. They seem to do the same thing,
but they don't have access to (or use) the same information:

- `default_signature` is used to construct type declarations from
  nothing; at this point, there is no immediacy information. On the
  other hand, it checks `Config.flat_float_array` and uses `Ind` instead
  of `Deepsep` when it is not set.

- `msig_of_external_type` takes the declaration of the type as
  argument. In particular, it checks immediacy (immediate types
  are always separable).

I thought about this, and I don't see any way to do better than having
these two functions. Note that `msig_of_external_type` is not exposed
directly by the `Typedecl_separability` signature, but it is accessible
as the `default` field of `Typedecl_separability.property`.

gasche commented:
(cc @damiendoligez, @yallop and @Armael; you might be interested in this PR)

Armael commented:
Don't count on me for this one, I won't have much time until a couple months.

yallop commented on typing/typedecl_separability.ml:
```diff
@@ -0,0 +1,668 @@
+let check_def
+  : Env.t -> type_definition -> Sep.signature
+  = fun env def ->
+  let unboxed = def.type_unboxed.unboxed in
```
Since every occurrence of `unboxed` is negated (`not unboxed`) I'd prefer to see this as
```ocaml
let boxed = not def.type_unboxed.unboxed
```
and have assertions without the double negations below:
```ocaml
assert boxed;
```

yallop commented on testsuite/tests/typing-unboxed-types/test_flat.ml:
```diff
@@ -193,10 +199,6 @@ type _ s = S : 'a t -> _ s  [@@unboxed]
-       it might contain both float and non-float values.
-       You should annotate it with [@@ocaml.boxed].
+type _ s = S : 'a t -> 'b s [@@unboxed]
+and _ t = T : 'a -> 'a s t
```
I'd like to see more examples of type definitions that were previously rejected but are now accepted.

For example, this example (from the paper) could be included:

```ocaml
type ('a, 'b) t = K : 'c -> ('c, 'c) t [@@unboxed]
         and t1 = T1 : ('a, int) t -> t1 [@@unboxed]
```

as could @mmottl's example from [Mantis 7364](https://caml.inria.fr/mantis/view.php?id=7364):

```ocaml
type ('a, 'kind) tree =
  | Root : { mutable value : 'a; mutable rank : int } -> ('a, [ `root ]) tree
  | Inner : { mutable parent : 'a node } -> ('a, [ `inner ]) tree

and 'a node = Node : ('a, _) tree -> 'a node [@@ocaml.unboxed]
```

gasche replied:
I added the two tests that you suggested -- we forgot to do it earlier -- and I changed some other tests to use named instead of perpetually unnamed existential variables. The testsuite still needs some of the more advanced cases that we found while preparing the paper (an example of non-principality, some of the trickier GADT cases, and some cases exercising the support for cyclic/equi-recursive types).

(So I'm not marking as "resolved" yet.)

gasche replied:
I added the tricky tests I had in mind, marking this resolved.

yallop commented on Changes:
```diff
@@ -613,6 +613,11 @@ Working version
+- GPR#2188: improvement of the unboxability check for types with a single
+  constructor. Mutually-recursive type declarations can now contain unboxed
+  types. This is based on the paper https://arxiv.org/abs/1811.02300.
+  (Gabriel Scherer and Rodolphe Lepigre, review by ?)
```
The changes entry should mention [Mantis 7364](https://caml.inria.fr/mantis/view.php?id=7364).

yallop commented on typing/types.mli:
```diff
@@ -286,6 +286,35 @@ module Variance : sig
+
+  type signature = t list
+  (** The 'separability signature' of a type assigns a mode for
+      each of its parameters. [('a, 'b) t] has mode [(m1, m2)] if,
```
stray comma after "if"

yallop commented on typing/types.mli:
```diff
@@ -286,6 +286,35 @@ module Variance : sig
+  (** Compare two mode according to their mode ordering. *)
+
+  val max : t -> t -> t
+  (** [max_mode m1 m2] returns the most demanding mode, it is used to
```
"mode, it" ↝ "mode. It"

yallop commented on typing/types.ml:
```diff
@@ -140,6 +140,22 @@ module Variance = struct
+    | Sep -> 1
+    | Deepsep -> 2
+  let compare m1 m2 = compare (rank m1) (rank m2)
+  let max m1 m2 = if rank m1 >= rank m2 then m1 else m2
```
I prefer 

```ocaml
let max m1 m2 = Pervasives.max (rank m1) (rank m2)
```

which makes it clearer that this is just the standard `max`, with nothing funny going on.

gasche replied:
But this would return the rank, not the mode itself. It would work if we had a `max_by`, but in the meantime I don't see a better way than what we have.

yallop replied:
Ah, true.  `Pervasives.max` would work, of course, but the code is better as it is.

yallop commented on typing/typedecl_separability.mli:
```diff
@@ -0,0 +1,125 @@
+    separability cannot be established. *)
+
+type mode = Types.Separability.t = Ind | Sep | Deepsep
+(** The mode [Sep] ("separable") is used on type expressions that must be
```
The comment phrases the mode as a requirement on arguments ("type expressions that must be separable"), and that's generally how it's used.  But I'd prefer to see it defined more neutrally somewhere as a description of type expressions.  One option is to document `Types.Separability.t` as a (representation of a) property of type expressions, and separately document `Typedecl_separability.mode` as a requirement on arguments, as it is currently.

gasche replied:
I rewrote this comment to include for each mode, then the type-expression view, and then the type-parameter view. I think the latter is necessary to give intuition (type expressions themselves, at the toplevel, must all be separable).

yallop commented on typing/typedecl_separability.ml:
```diff
@@ -0,0 +1,668 @@
+
+(** assuming that a datatype has a single constructor/label with
+   a single argument, [argument_to_unbox] represents the
+   information we need to check a the argument for separability. *)
```
"check a the"

gasche commented:
@yallop: thanks! I fixed the easy issues, changes for the other are yet to come.

Commit:
typedecl_separability: interfaces and basic definitions, implementations missing

gasche committed:
bootstrap

gasche committed:
typedecl_separability: support for GADTs

gasche committed:
typedecl_separability: protect against infinite cyclic types

Commit:
typedecl: use the new Separability implementation, remove old code

Because this changes the separability of standard library types, and
those separabilities are stored in the .cmi files, this commit changes
the .cmi files in the standard libraries in way that appear to require
a bootstrap (it looks like some part of the stdlib is built with
boot/ocamlc and others with ocamlc, and the two should produce/expect
the same .cmi exactly). The bootstrap will come as a separate commit.

gasche committed:
separability: add some interesting examples to the testsuite

gasche commented:
I rebased the PR history, it reads like a novel now. (cc @rlepigre)

gasche committed:
bootstrap

rlepigre commented:
@gasche: A bestseller for sure!

gasche commented:
I rebased against the current trunk and finished taking into accounts the comments of @yallop's first review pass. Thanks!

gasche commented:
(cc @let-def, who seemed interested in our presentation of this work at JFLA 2019.)

(Jeremy, it's not clear to me whether you are satisfied with the PR as it is proposed.)

gasche commented:
There is nothing very urgent on this PR (... well it does fix a limitation with an existing feature), but it would still be nice to find someone interested in making a full review. Maybe someone among @yallop, @Armael, @Octachron, @Drup, @let-def, @trefis?

damiendoligez commented on typing/typedecl_separability.ml:
```diff
@@ -0,0 +1,674 @@
+exception Error of Location.t * error
+(*
+exception Not_separable of { loc: Location.t; evar: string option; }
+*)
```
This looks like old code that was commented out.

damiendoligez commented on typing/typedecl_separability.ml:
```diff
@@ -0,0 +1,674 @@
+      immediate_subtypes_variant_row [] row
+
+  (* the cases below are not called from check_type_expr,
+     they are here for completeness *)
```
I'm a bit meh about including the cases below: if the comment is correct they are dead code.

gasche replied:
(I'm not fully sure that the comment is correct.) When we split some logic from a specific site into a generic function, it often happens that the natural specification of the generic function covers more than the interface used at the callsite. Either we make the specification more complex by ruling certain cases out, or we write dead code. In OCaml the former is made by adding `assert false`. The difference will be noticeable when someone uses the generic function in other places, or changes the caller in a way that invalidates the restrictions there; if you wrote `assert false`, then you will now blow up in some cases, if you handled these cases you run the risk of bugs because those paths were not tested. I would prefer the second alternative, relying on my great reviewers to also double-check that those cases make sense.

damiendoligez commented on typing/typedecl_separability.ml:
```diff
@@ -0,0 +1,674 @@
+      (* fold_right here is necessary to get the mode
+         consed to the accumulator in the right order *)
+      List.fold_right handle_equation parameters ([], context) in
+    (* After all variable determined by the parameters have been set to Ind
```
```suggestion
    (* After all variables determined by the parameters have been set to Ind
```

damiendoligez commented on typing/typedecl_separability.ml:
```diff
@@ -0,0 +1,674 @@
+
+let property : (prop, unit) Typedecl_properties.property =
+  let open Typedecl_properties in
+  let eq ts1 ts2 = try List.for_all2 Sep.eq ts1 ts2 with _ -> false in
```
```suggestion
  let eq ts1 ts2 = try List.for_all2 Sep.eq ts1 ts2 with Invalid_argument _ -> false in
```

`try ... with _ ` is bad, see the whole discussion about asynchronous exceptions.

damiendoligez commented on typing/typedecl_separability.mli:
```diff
@@ -0,0 +1,131 @@
+
+    (Note: This assumption is required for the dynamic float array optimization;
+    it is only made if Config.flat_float_array is set,
+    otherwise the code in this module defaults becomes trivial
```
```suggestion
    otherwise the code in this module becomes trivial
```

damiendoligez commented on typing/typedecl_separability.ml:
```diff
@@ -0,0 +1,674 @@
+let property : (prop, unit) Typedecl_properties.property =
+  let open Typedecl_properties in
+  let eq ts1 ts2 = try List.for_all2 Sep.eq ts1 ts2 with _ -> false in
+  let merge ~prop:_ ~new_prop = new_prop in
```
Could you add a comment to explain why `merge` just ignores its first argument?

damiendoligez reviewed (commented):
Looks good except for a few nitpicks.

let-def commented on typing/typedecl_separability.mli:
```diff
@@ -0,0 +1,131 @@
+   similar to what is done for inferring the variance of type parameters.
+
+   Our analysis is defined using inference rules "Def; Gamma |- t : m", in
+   which a type expression "t" is checked agains a "mode" "m". This "mode"
```
s/agains/against

let-def commented on typing/typedecl_separability.ml:
```diff
@@ -0,0 +1,674 @@
+
+   Note: A GADT constructor introduces existential type variables, and
+   may also introduce some equalities between the type parameters
+   (or universal type variables) of the type it is constructed and
```
"...of the type it is constructed and..."
I don't understand this sentence, I think a word is missing or wrong.
"the type that is constructed", "being constructed", ...?

gasche replied:
I wasn't sure either what we meant there anymore. I rephrased to something I understand. Thanks!

let-def reviewed (commented):
I am not done reviewing this PR, but I already have a few minor remarks.

gasche committed:
changes from review comments

gasche commented:
I am in the progress of rebasing this PR on trunk (it requires a bit of work). Please keep sending the good review comments in the meantime :-)

rlepigre commented:
> I am in the progress of rebasing this PR on trunk (it requires a bit of work). Please keep sending the good review comments in the meantime :-)

Thanks for taking care of that @gasche. I don't have much time currently, but is there anything I can help with?

gasche commented:
I just finished rebasing on trunk and pushed to this branch.

@rlepigre: Thanks! I'm happy with dealing with the mostly local/form comments so far, but we can definitely discuss it if more ambitious changes are requested by reviewers.

smuenzel-js commented:
how's this PR going?

damiendoligez commented on typing/typedecl_separability.ml:
```diff
@@ -0,0 +1,675 @@
+    | (Tlink _            , _      ) -> assert false
+    (* Impossible case (according to comment in [typing/types.mli]. *)
+    | (Tsubst(_)          , _      ) -> assert false
+    (* "Indiferent" case, nothing to do. *)
```
```suggestion
    (* "Indifferent" case, the empty context is sufficient. *)
```

damiendoligez commented on typing/typedecl_separability.ml:
```diff
@@ -0,0 +1,675 @@
+    | (Tpackage(_,_,_)    , Deepsep) ->
+        let tys = immediate_subtypes ty in
+        let on_subtype context ty =
+          context ++ check_type (Hyps.guard hyps) ty (compose m Ind) in
```
```suggestion
          context ++ check_type (Hyps.guard hyps) ty Deepsep in
```
We know that `m` is `Deepsep`, no need to call `compose`. Also, I think it's clearer this way because for something to be Deepsep, we simply need all its arguments to be Deepsep.

gasche committed:
typedecl_separability: handle GADT equations in a more predictable order

The separability signature of a type declaration is not inferred in
a principal way, it depends on the order in which GADT equations are
processed. In non-principal cases, we may have two parameters that are
related by an equality, with one of them being given mode Ind and the
other Sep. Either choice of which to make Sep is sound, but (Ind, Ind)
would be unsound.

We change the implementation to ensure that equations are processed in
an order such that the lefmost parameters are the most constained: if
equations imply that ('a = 'b), with the parameter 'a coming before 'b
in the type declaration, and they must be separable, then 'a gets the
mode Sep and 'b gets the mode Ind. This corresponds intuitively to
remembering that 'b is equal to a previous parameter, instead of
remembering than 'a is equal to a not-seen-yet parameter.

gasche commented:
After @smuenzel-js's ping, @damiendoligez started to review again (I think he hasn't validated his last review so his comments do not show up). We discussed the non-principality issue together (to me this is the last standing issue that required a discussion).

Non-principality shows up for types such as

```ocaml
type ('a, 'b) almost_eq =
| Almost_refl : 'c -> ('c, 'c) almost_eq
```

with (at least) two parameters `('a, 'b)` which are in fact always equal, and must be separable. We can give the separability modes `(Sep, Ind)` or also `(Ind, Sep)`, which allow more programs thant `(Sep, Sep)`, but `(Ind, Ind)` would be unsound.

We decided that there was, in fact, a more natural choice among the two options: it is more natural to constrain the first/leftmost type parameter more, and to let the second be `Ind` because it is equal to the first.

I just pushed a last commit to the PR that enforces this ordering, by processing GADT equations from left to right.

damiendoligez commented on typing/typedecl_separability.mli:
```diff
@@ -0,0 +1,131 @@
+   In general, the analysis performs a fixpoint computation. It is somewhat
+   similar to what is done for inferring the variance of type parameters.
+
+   Our analysis is defined using inference rules "Def; Gamma |- t : m", in
```
Strictly speaking, `Def; Gamma |- t : m` is a statement, not an inference rule.

damiendoligez reviewed (approved):
Looks good to me, modulo a few nitpicks.

gasche commented:
@damiendoligez thanks, I took your suggestions into account and updated the branch.

gasche commented:
I just rebased on trunk again, this should be good to merge once the CI is happy.

shindere commented:
It seems that commit 358c7cecd5cbd923742f80fabc8e6fc41af26955 breaks trunk
when configured with the
`--disable-flat-float-array` option.

The problem is very easy to reproduce. Checkkout the commit mentionned
above, git clean -xffd, configure with the `--disable-flat-float-array`
option and run make. It should lead to inconsistent assumptions over an
interface.

@gasche do you think you'd be able to fix this?

FWIW a job on Inria's CI had detected the problem but it seems it didn't
receive much attention.

gasche commented:
On my machine, the current HEAD of `trunk` builds correctly with `--disable-flat-float-array`, and the testsuite passes. The commit you point to gives a disfunctional state because it requires a bootstrap and we have a not-so-good policy to put bootstraps in separate commit.

shindere commented:
@gasche can you please have a look to the other-configs job on Inria's CI?

gasche commented:
So there is indeed a CI bug when `--disable-flat-float-array` and `--enable-flambda` are passed at the same time, which I could reproduce on my machine. It looks like a Makefile dependency ordering issue, because I could get the build to work by cleaning up some bits manually and redoing the rest of the build. I'm not sure why only this configuration is affected, but note that this PR changes `.cmi` format and their digests in ways that required bootstraps, so that may explain why certain cmi-digest issues are now showing up.

gasche commented:
I'm starting to understand what is going on, writing it down here.

The first thing to understand is that different `cmi` files get produced by compilers depending on whether `--disable-flat-float-array` is set: a compiled interface exports the "separability" of each datatype definition, and when flat-float-array is disabled all datatypes are separable, so the information is different (more permissive) from when it is not set.

In the case where `--disable-flat-float-array` is set, the bootstrap compiler comes from a bootstrap *without* this setting, so `boot/ocamlc` and `ocamlc` generate different `cmi`. (They have the same format and, in this setting, compiling with *either* cmis is sound/correct; the bootstrap-compiler-produced cmis are more restrictive, but we know the compiler codebase passes them anyway as they are the standard mode.)

Now here is a partial analysis of the (failing) build logs:

```
$ grep -E "( map.ml|/misc.ml)" log
../boot/ocamlrun ../boot/ocamlc -use-prims ../runtime/primitives -strict-sequence -absname -w +a-4-9-41-42-44-45-48 -g -warn-error A -bin-annot -nostdlib -safe-string -strict-formats   -o stdlib__map.cmi -c map.mli
../boot/ocamlrun ../boot/ocamlc -use-prims ../runtime/primitives -strict-sequence -absname -w +a-4-9-41-42-44-45-48 -g -warn-error A -bin-annot -nostdlib -safe-string -strict-formats   -o stdlib__map.cmo -c map.ml
./boot/ocamlrun ./boot/ocamlc -g -nostdlib -I boot -use-prims runtime/primitives -strict-sequence -principal -absname -w +a-4-9-40-41-42-44-45-48-66 -warn-error A -bin-annot -safe-string -strict-formats -I utils -I parsing -I typing -I bytecomp -I file_formats -I lambda -I middle_end -I middle_end/closure -I middle_end/flambda -I middle_end/flambda/base_types -I asmcomp -I asmcomp/debug -I driver -I toplevel -c utils/misc.mli
./boot/ocamlrun ./boot/ocamlc -g -nostdlib -I boot -use-prims runtime/primitives -strict-sequence -principal -absname -w +a-4-9-40-41-42-44-45-48-66 -warn-error A -bin-annot -safe-string -strict-formats -I utils -I parsing -I typing -I bytecomp -I file_formats -I lambda -I middle_end -I middle_end/closure -I middle_end/flambda -I middle_end/flambda/base_types -I asmcomp -I asmcomp/debug -I driver -I toplevel -c utils/misc.ml
../boot/ocamlrun ../ocamlc -strict-sequence -absname -w +a-4-9-41-42-44-45-48 -g -warn-error A -bin-annot -nostdlib -safe-string -strict-formats   -o stdlib__map.cmi -c map.mli
../boot/ocamlrun ../ocamlc -strict-sequence -absname -w +a-4-9-41-42-44-45-48 -g -warn-error A -bin-annot -nostdlib -safe-string -strict-formats   -o stdlib__map.cmo -c map.ml
           -o stdlib__map.cmx -c map.ml
./boot/ocamlrun ./ocamlopt -g -nostdlib -I stdlib -I otherlibs/dynlink -strict-sequence -principal -absname -w +a-4-9-40-41-42-44-45-48-66 -warn-error A -bin-annot -safe-string -strict-formats -I utils -I parsing -I typing -I bytecomp -I file_formats -I lambda -I middle_end -I middle_end/closure -I middle_end/flambda -I middle_end/flambda/base_types -I asmcomp -I asmcomp/debug -I driver -I toplevel -function-sections -c utils/misc.ml
```

What I see is that:
1. `stdlib/map.{mli,ml}` is built with boot/ocamlc
2. `utils/misc.{mli.ml}` is built with boot/ocamlc
3. `stdlib/map.{mli,ml}` is rebuilt with ocamlc, generating different cmi files
4. `utils/misc.ml` is rebuilt with ocamlopt to get a .cmx, and there is a cmi mismatch

At step (4), `utils/misc.ml` is rebuilt (with ocamlopt, but ocamlc would mismatch as well), but `misc.mli` is not rebuilt. This may sound reasonable as the textual dependencies of `misc.{mli,ml}` have not changed, but it is in fact incorrect as the previous build happened with the bootstrap-produced stdlib cmis, and we are now building against the normal-compiler-produced stdlib.

I don't think that this PR is to blame for this problem, it's rather one of the rare few PRs doing this kind of things (changing cmis), in a funny enough way (depending on configure flags) that it caught the issue. There is missing dependency information in the Makefile, which should notice that utils/misc.mli needs rebuilding as stdlib/map.cmi changed.

gasche commented:
I created a new issue, #9291, to discuss the build failure and Makefile dependency issues. I will not discuss it more on the present thread.
